Ft user password reset 165742482

sudhons opened the pull request:
# Description

`Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.`

# Issue ID/Link

`Choose one below and replace with the relevant Issue ID or Link. Delete the unused option`

- [ ] Pivotal Tracker [#Issue Id](https://www.pivotaltracker.com/story/show/<IssueId>)
- [ ] Jira [#Issue Id](https://www.jira.com/story/show/<IssueId>)

# What kinds of changes does this PR introduce?

`Please delete options that are not relevant.`

- [ ] Non-breaking **_[Bug Fix]_** which fixes an issue/bug.
- [ ] Non-breaking **_[Feature]_** which adds functionality.
- [ ] Non-breaking **_[Chore]_** which adds functionality.
- [ ] Breaking change **_[Fix|Feature|Chore]_** that would/may cause existing functionality not to work as expected.
- [ ] This change requires a documentation update.

# How Has This Been Manually Tested?

`Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce and/or test it accordingly.`

- Scenaria for Test A
- Scenario for Test B etc...

# Test Coverage:

`Add test coverage details for before and after.`

- Current Coverage: []%
- New Coverage if merged: []%

# Checklist:

`Ensure you are able to check all the listed options below. If not fill the additional context in the section below.`

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# Additional Context

`Optional section, can be used to give insights, or just lodge complaints regarding this PR.`

henperi commented on src/controllers/AuthController.ts:
```diff
@@ -65,6 +72,129 @@ class AuthController {
+        });
+      }
+
+      res.locals = { ...res.locals, dbUser: user };
```
Can we just have our user in `res.locals.user `as opposed to `res.locals.dbUser`?

henperi commented on src/controllers/AuthController.ts:
```diff
@@ -65,6 +72,129 @@ class AuthController {
+  static async updatePassword(req: Express.Request, res: Express.Response): Promise<void> {
+    try {
+      const user = await UserRepository.updatePassword(
+        res.locals.dbUser,
```
consider `res.locals.user` as opposed to `req.locals.dbUser`

henperi commented on src/helpers/tokenHelpers.ts:
```diff
@@ -22,6 +22,30 @@ const generateUserToken = (userTokenData: TokenData): string => {
+const generatePasswordResetToken = (tokenData: Partial<TokenData>): string => {
+  try {
+    return jwt.sign(tokenData, PASSWORD_RESET_SECRET, {
+      expiresIn: '1d',
```
I think we should go for 3 hours or less here.

henperi commented on src/routes/auth/index.ts:
```diff
@@ -13,4 +14,16 @@ const authRouter = express.Router();
 authRouter.post('/signup', validateCreateUser, validatePhoneNumber, AuthController.createUser);
 
+authRouter.post(
+  '/reset-password',
```
Also, consider changing the endpoint name to something that relates to sending the reset email.

henperi commented on src/controllers/AuthController.ts:
```diff
@@ -65,6 +72,129 @@ class AuthController {
+   * @param res express response
+   * @returns Promise<void>
+   */
+  static async resetPassword(req: Express.Request, res: Express.Response): Promise<void> {
```
Great implementation here bro, but I'm looking at the name of this method(resetPassword) and the description(Method to reset a User's password) above it. I think we should modify its name/description.
*Why?* `Because this method doesn't really reset the password. It just sends an email that helps you reset the password.`

I think the name should be related to sending reset email or something in that line. Perhaps `sendResetPasswordEmail`, or `resetPasswordRequest` might suffice.

I'd like to know your thoughts on my perspective.

sudhons committed:
feat(password-reset): password reset functionality
- add password reset request endpoint
- add update endpoint
- write unit tests
- write integration tests
[#165742482]

henperi commented on package.json:
```diff
@@ -85,6 +86,7 @@
     "@types/faker": "^4.1.9",
     "@types/google-libphonenumber": "^7.4.17",
     "@types/sequelize": "^4.28.5",
+    "@types/sinon": "^7.5.1",
```
This should be in dev dependecy

henperi commented on src/controllers/AuthController.ts:
```diff
@@ -65,6 +72,121 @@ class AuthController {
+   * @param next express function
+   * @returns Promise<void>
+   */
+  static async checkUser(
```
This middleware method should be in the middleware folder

henperi commented on tests/unitTests/controllers/AuthController.spec.ts:
```diff
@@ -0,0 +1,158 @@
+      sinon.replace(AppResponse, 'success', success);
+      sinon.replace(tokenHelpers, 'generatePasswordResetToken', generatePasswordResetToken);
+
+      await AuthController.handlePasswordRestRequest(request, response);
```
The name `handlePasswordRestRequest` should be modified to `handlePasswordResetRequest`

henperi commented on src/controllers/AuthController.ts:
```diff
@@ -6,8 +6,15 @@ import UserRepository from '../repositories/UserRepository';
+  validatePasswordResetToken,
+} from '../helpers/tokenHelpers';
+
+const passwordRestCallbackURL = process.env.PASSWORD_RESET_CALLBACK_URL;
```
I noticed dotenv is not imported in this file. Are we still able to access environment variables like this?

sudhons replied:
You don't need `dotenv` to access environment variables from `process.env`. You only need it to load the variables from `.env` file into `process.env` and you can always access it from there.
You import `dontenv` once in your application that's why it's important to do this early in your application before any file needs any environment variable.

henperi commented on src/controllers/AuthController.ts:
```diff
@@ -65,6 +72,121 @@ class AuthController {
+   * @param res express response
+   * @returns Promise<void>
+   */
+  static async handlePasswordRestRequest(
```
Theres a typo on this method's name. I think you meant to write `handlePasswordResetRequest `

henperi commented:
Kindly rebase as well
